rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeper.java

```diff
@@ -585,27 +617,37 @@ public void onRemoval(RemovalNotification<String, FileMetadata> notification)
     Configuration conf;
     long fileLength;
     long lastModified;
+    long currentFileSize;
 
-    public CreateFileMetadataCallable(String path, long fileLength, long lastModified, Configuration conf)
+    public CreateFileMetadataCallable(String path, long fileLength, long lastModified, long currentFileSize,
+                                      Configuration conf)
     {
       this.path = path;
       this.conf = conf;
       this.fileLength = fileLength;
       this.lastModified = lastModified;
+      this.currentFileSize = currentFileSize;
     }
 
     public FileMetadata call()
         throws Exception
     {
-      return new FileMetadata(path, fileLength, lastModified, conf);
+      return new FileMetadata(path, fileLength, lastModified, currentFileSize, conf);
     }
   }
 
-  public static void invalidate(String p)
+  // This method is to invalidate FileMetadata from guava cache.
+  // deleteCachedFile determines whether to delete the actual file from the local filesystem or not
+  private static void invalidateFileMetadata(String key, boolean deleteCachedFile)
   {
     // We might come in here with cache not initialized e.g. fs.create
     if (fileMetadataCache != null) {
-      fileMetadataCache.invalidate(p);
+      FileMetadata metadata = fileMetadataCache.getIfPresent(key);
+      fileMetadataCache.invalidate(key);
+
+      if (metadata != null && deleteCachedFile) {
```

Comment:
I have put the same key and a new object of FileMetadata (that will take care of the weight). This will cause the older value to be REPLACEd. We will get Removal Notification with cause REPLACE at that time. 

So Invalidate is going to evict the file with cause EXPLICIT  and we will delete the files if the cause is not REPLACE

Reply:
👍